### ISSUE TITLE
redux-epic and redux-observable

### ISSUE THREAD

**Issue by BerkeleyTrue** [redux-observable/redux-observable#84]
Hello Everyone,

I'm the creator of Redux-Epic, which is a redux library in similar fashion to redux-saga but built using RxJs with first class support for SSR.

In https://github.com/BerkeleyTrue/redux-epic/issues/13 it has been brought to my attention that this library has arrived at the same path as redux-epic, triggering observable in middleware a la redux-saga.

But there are also various differences that I go into detail in the issue above.

I'm opening this issue to discuss whether merging the two products would benefit everybody and whether the goals I've laid out for redux-epic are compatible with the goals of redux-observable.
- First class support for SSR
- Designed for Universal JavaScript (read: Isomorphic)
- Ability to inject dependencies
- Disposable Epics

**Comment by jayphelps** [comment]
Nice! I'll provide some of my initial thoughts.

Because we both use the term Epic, but seem to use them differently, for the purpose of the following Epic will **always** mean redux-observable's concept of Epic, which I believe is still called Saga in redux-epic?

### First Class Support for Server Side Rendering
We'd love this, but have complex concerns. I can't immediately tell if they are solved by redux-epic.

Summary: We will have to spend more time around this. Our concerns may end up unfounded.

### Universal JavaScript First Design
I think this is the same thing as First Class Support for Server Side Rendering?

### Dependency injection
Looking at the docs, I'm not sure how this is implemented in redux-epic so I can't comment on it. I imagine it's similar to being able to inject a third param to all Epics. Often an API library or something, so they can easily mock it in testing.

```js
const fetchUserEpic = (action$, store, api) =>
  action$.ofType(FETCH_USER).etc();
```
We've been experimenting with exactly this internally. If it continues to work well, we do plan to add it or obviously the community could PR this ability before hand.

This is a pretty obvious solution, so I imagine it's very similar to the one you landed one too.

Summary: We are experimenting with patterns that may eventually become part of the library to try and minimize API churn.

### RxJS v4 instead of RxJS v5
### Once RxJS is out of beta and has fully replaced Rx then this library will move over.
I believe this one is the Number 1 practical difference between the two, in my opinion.

[The goals of v5 are very important to me](https://github.com/ReactiveX/rxjs#goals), though I imagine not for everyone. I work at Netflix, so there's clearly going to be some inherent bias. There are the [code, performance, debugging, etc](https://github.com/ReactiveX/rxjs#goals) (basically DX) differences but as you mentioned also the bureaucratic differences. RxJS v5 _will replace v4_. The v5 effort was initiated and mostly been driven by Netflix, but is indeed [a collaboration with Microsoft](https://github.com/ReactiveX/rxjs/blob/master/LICENSE.txt#L189). They've helped make some of the core decisions and changes. That said, v4 will likely still be maintained for some time because of its proliferation and relative stability.

I can absolutely see why someone would choose v4 over v5 believe they want code with a longer track record. v5 still has a beta tag not because we believe it is not production ready, but because there are some [breaking changes that are being wrapped up](https://github.com/ReactiveX/rxjs/labels/BLOCKING-FIVE-OH-RELEASE) that really are fairly minor in dev impact, but non-trivial for the maintainers. e.g. browser related APIs being extracted to a separate `rxjs-dom` library via a monorepo.

Summary: I can safely say that redux-observable will stay RxJS v5 internally with the possibility of supporting other Observable implementations at a later date like v4, most.js, bacon, etc. The biggest thing is keeping primary support for the proposed TC39 Observable spec.

### Disposable Epics
We intentionally chose to keep a simple API design initially because it's fairly easy to create your own patterns on top of our Epics, since they are just a regular function that expects a stream of actions and returns a stream of actions. That means you can indeed dispose our Epics, just like you'd dispose any other observable. We don't provide any explicit means for this, because it's just RxJS. 😄

Summary: We do want to get better docs around this, even if they're mostly just repeating regular RxJS patterns in the context of redux-observable. If we provided any abstractions, it would probably be around the root Epic in the middleware itself for SSR.

### React Centric
While we do plan to add more contributors from the community who shine, currently redux-observable is maintained by myself and @blesh. We're both primarily React people _right now_ but we've changed UI frameworks so many times that we didn't want to make this library React specific. We wanted to focus on the core concepts of the library first, then as we became more confident in them we'd extract common patterns for specific UI frameworks, into their own libraries. e.g. react-redux-observable, ember-redux-observable, angular-redux-observable, etc. We also thought the community might step in first and create their own that are even better than the ones we'd create.

Summary: We're hoping to get some core contributors from other UI communities like Angular2/Ember and want to keep redux-observable agnostic, differing to any future addon libraries to provide UI framework sugar.

### Joining Forces
Totally would love this, if possible. Not sure how we would go about it.

**Comment by SethDavenport** [comment]
For the 'react-centric' point, here are some opinions from the [Angular 2 side of the table](https://github.com/angular-redux/ng2-redux/blob/master/docs/epics.md).

:)

My own self-serving opinion is that I like that this lib is _not_ react centric.

Also, Angular 2 uses RxJS 5 observables extensively, so pulling in v4 too would be awkward.

Finally, when it comes to DI, I'm not opposed to it but I would like it to be optional - Angular 2 already has its own DI solution.

**Comment by BerkeleyTrue** [comment]
@jayphelps Thanks for the response. I think we are in agreement on most points.

Also, I'm definitely not implying that this library should support Rx@4 or that it should be react centric, I am merely pointing out the differences and my reasoning behind the choices I made with Redux-Epic.

# Disposable Epics

> it's fairly easy to create your own patterns on top of our Epics, since they are just a regular function that expects a stream of actions and returns a stream of actions. 

I think there is some confusion around the term `epic` here because it sounds like you are discussing the user created saga/epic while I'm refer to the library created middleware `epic`. (I'm have trouble keeping track as well :))

My main concern with SSR and this library is this [line](https://github.com/redux-observable/redux-observable/blob/master/src/createEpicMiddleware.js#L19). It seems to be the main point where the epic is called and the returned observable is subscribed to.

The subscription here is not reachable by a user. As a user, how would I dispose of this subscription? If I'm doing SSR, I'm creating an epicMiddleware per request on my server. Does this subscription hang around after the request has completed? 

Otherwise I think there is a lot common ground and making this library support SSR shouldn't involve too many changes. I'm also liking the `ofType` operator. I might implement something similar as a helper function.

**Comment by jayphelps** [comment]
All--we're going to discuss this between us in a Hangout, I'll try and distill the gist of the convo here after.

@BerkeleyTrue 

(FYI last repeat of the disclaimer that I will use the term Epic to refer to redux-observable's implementation 💃)

re: subscription cancelling, I think an example will clear up what I meant by "just use RxJS":

```js
const rootEpic = (action$, store) => {
  const out$ = new Rx.Subject();
  const subscription = combineEpics(...epics)(action$, store)
    .subscribe(out$);
  
  // You now control the subscription to the Epic.
  subscription.unsubscribe();

  return out$;
};
```
Also--we support `epicMiddleware.replaceEpic(nextEpic)` which lets you (as it sounds) replace the epic receiving actions. This is one way to "control" the subscription because we use `switchMap`, as you linked to. The old Epic will shutdown synchronously.

In the SSR world however, we haven't narrowed down the best way to "know" when all your Epics are done for that request and the final version of the page can be rendered and sent. Seemingly the solution might be to have the input Subject for actions `complete()` (`onComplete()` for you v4 peeps), so that the Epic can finish any async work it's doing but will otherwise complete itself once done. However, this also means that actions that Epics emit async won't be received by any other Epics either..which is a possible concern, even though likely an edge case in practice. Perhaps you have an elegant solution around this?

**Comment by BerkeleyTrue** [comment]
I see now! We could probably release a companion library to this one, say redux-observable-ssr, that does just that above and provide the renderToString method.

**Comment by jayphelps** [comment]
### Summary of Hangout between @BerkeleyTrue @blesh @jayphelps

We are aligned on long term goals. Short term, @BerkeleyTrue needs to continue to support RxJS v4 until he's confident they can switch to v5 and they have the cycles to do so. To facilitate this transition for v4 users, I'm experimenting with creating a compatibility suite:

1. a library that adds operators to both the v5 and v4 Observables for converting between the two:

  ```js
  const fetchUserEpic = action$ =>
    action$.v4().anyV4Operators().v5();
  ```

2. Adding an adapter interface for the Epic middleware, so you can convert to v4 and back to v5 automatically for all Epics, if you choose.

  ```js
  const epicMiddleware = createEpicMiddleware(rootEpic, { adapter: v4Adapter });
  ```

I've actually already written working version of both of these, just need to make the time this weekend or next week to publish them and add support in redux-observable. Here's the gist:

```js
function toV5(input$) {
  if (input$[$$observable]) {
    return input$[$$observable];
  } else {
    return ObservableV5.create(function (observer) {
      var subscription = input$.subscribe(
        observer.next.bind(observer),
        observer.error.bind(observer),
        observer.complete.bind(observer)
      );

      return function () {
        return subscription.dispose();
      };
    });
  }
}

function toV4(input$) {
  // Only v4 and below has this
  if ('flatMapLatest' in input$) {
    return input$;
  } else {
    return ObservableV4.create(function (observer) {
      var subscription = input$.subscribe(
        observer.onNext.bind(observer),
        observer.onError.bind(observer),
        observer.onCompleted.bind(observer)
      );

      return function () {
        return subscription.unsubscribe();
      };
    });
  }
}

const v4Adapter = {
  input: toV4,
  output: toV5
};
```

If this solution ends up not being friendly, in the short term @BerkeleyTrue is going to work on moving redux-epic to basically be a clone of redux-observable, but supporting RxJS v4. The other features it contains, like the React specific SSR stuff, can move to another library.

**Comment by jayphelps** [comment]
I've published an initial version of an interop library between RxJS v4 and v5: https://github.com/jayphelps/rxjs-compatibility

Will work on the adapter stuff for redux-observable next for automatic conversion between the two, but people are welcome to install rxjs-compatibility and use this now like this until then:

```js
const fetchUserEpic = actions$ =>
  action$.ofType(FETCH_USER)
    .v4() // convert to v4
    .anyV4OperatorsYouWant()
    .v5(); // convert back to v5
```

Since RxJS v5 does _not_ import any operators by default, this really isn't a huge deal from a bundling size standpoint as redux-observable bundles the `.ofType()` and its own operator dependencies. That is, you're not bundling all of v5 when you use it like this.

**Comment by jayphelps** [comment]
Initial version of adapters PR is here #85 and the v4 adapter [here](https://github.com/redux-observable/redux-observable-adapter-rxjs-v4)

**Comment by robmcm** [comment]
Has there been any progress on dependency injection (additional arguments) as mentioned by @jayphelps ?

**Comment by BerkeleyTrue** [comment]
@robmcm Additional arguments are already possible.

**Comment by robmcm** [comment]
Thanks @BerkeleyTrue but I can't see anything in the docs (https://redux-observable.js.org/?q=Additional%20arguments) or in this repo, could you point me in the right direction?

**Comment by BerkeleyTrue** [comment]
Epics are plain functions. You can wrap your root epic in a closure and then pass that to the epic middleware creator. 

https://github.com/realworldreact/react-shoppe/blob/intro-rx/common/create-store.js#L19

**Comment by jayphelps** [comment]
@robmcm we need to update the docs. One way is like this:

```js
const somethingEpic = (action$, store, { api }) =>
  action$.ofType(SOMETHING)
    .mergeMap(action =>
      api.fetchSomething(action.payload)
        .map(response => somethingFulfilled(response))
    );

// when defining the rootEpic, inject deps
import * as api from 'wherever';

const rootEpic = (action$, store) =>
  combineEpics(...epics)(action$, store, { api })

// later, to test...

const action$ = ActionsObservable.of({ type: SOMETHING, payload: '123' });
const response = { id: '123', something: 'example' };
const api = { fetchSomething: id => Observable.of(response) };

somethingEpic(action$, null, { api })
  .toArray()
  .subscribe(actions =>
    assertDeepEqual(actions, [somethingFulfilled(response)]);
  );
```

**Comment by robmcm** [comment]
Thanks @jayphelps 

I think it would be a great addition to the docs. 

On a side note, it would be also nice to add an example of cancelling an infinite stream with the same event that triggers it. Say a stream that ticks ever n seconds, where n comes from an input. The concept of teardown, build up from the same action isn't clear to me yet. 

I am relatively new to RX so it may be that this is more applicible in those docs, please ignore if that's the case.

**Comment by jayphelps** [comment]
@robmcm 

> On a side note, it would be also nice to add an example of cancelling an infinite stream with the same event that triggers it. Say a stream that ticks ever n seconds, where n comes from an input. The concept of teardown, build up from the same action isn't clear to me yet.

I'm not entirely sure what you mean.

**Comment by robmcm** [comment]
@jayphelps 

I'm the cancellable recipe the example requires a cancel action (coming from a cancel button). This is simple but not a very common UX pattern. 

What if the username was entered into a text box and the results were streamed into the view? In this case I would imagine the username change action would need to trigger a new lookup and cancel any existing one. 

The solution may be identical/trivial but it's something that I am currently unsure of and need to look into.

**Comment by jayphelps** [comment]
@robmcm the cancel example uses a button simply for the demo, but usually the cancel action is actually dispatched automatically, e.g. in `componentWillUnMount`

> What if the username was entered into a text box and the results were streamed into the view? In this case I would imagine the username change action would need to trigger a new lookup and cancel any existing one.

This sounds like a case for [`switchMap`](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-switchMap) instead of using `mergeMap`:

```js
const fetchUserEpic = action$ =>
  action$.ofType(FETCH_USER)
    .switchMap(action => // <-- switchMap, not mergeMap
      ajax.getJSON(`/api/users/${action.payload}`)
        .map(response => fetchUserFulfilled(response))
    );
```

Which will have the effect of cancelling any pending AJAX request when another request comes in before it finishes.

**Comment by robmcm** [comment]
@jayphelps

Thanks, I guess a lot of these things are just implementation details (perhaps more obvious to an RX native).

**Comment by SethDavenport** [comment]
@robmcm Possibly off topic but since Angular kind of makes you use DI we do it like this https://github.com/angular-redux/ng2-redux/blob/master/docs/epics.md

Essentially this is an example of 'using a closure' as suggested above: Angular's HTTP module is saved in the instance's closure and as accessible by the epic. A bit messier because Angular kind of forces us into classes but it works.

**Comment by robmcm** [comment]
@SethDavenport That's really helpful actually, and similar to my quite hack for our POC. We are using TS so it's useful to see how you are using it with redux-observable and RX in general.

**Comment by jayphelps** [comment]
@SethDavenport using classes and formal DI is more work--but it does make it more future proof and "correct" in the sense that you're not overloading the argument signature of your epics. This fact is why I've been hesitant to officially document the regular, non-class DI pattern, because if we ever want to add arguments to epics, it will break things. e.g. we've talked about a signature of `(action$, state$, store)` but having used redux-observable for almost a year and helped countless people, I've yet to come across a real world use case for state as a stream. And if someone really wants it, they can DI it just like any other deps.

**Comment by stevematdavies** [comment]
I simply cant get this to work. (It works using Axios and similar)

const loginEpic = (actions) =>    
    ```
 actions.ofType(ActionTypes.LOGIN_REQUESTED)
        .switchMap(({credentials}) =>
            Observable
                .ajax.getJSON('http://localhost:3001/login')
                .map((loginResponse) => 
                    loginResponse.status === 200
                        ? loginResolvedAction(loginResponse.response)
                        : loginFailedAction({message: 'Error'})));
```

Probably going to avoid epics this way until some kind soul can provide documentation or examples.

**Comment by stevematdavies** [comment]
I simply cant get this to work. (It works using Axios and similar) Ive avoided attempting to pass in any credentials to the request at the moment, since it appears rxjs can't handle them (well no docs or examples exist, illustrating how to pass params to the ajax get request (other than exposing them in the url))

const loginEpic = (actions) =>    

 actions.ofType(ActionTypes.LOGIN_REQUESTED)
        .switchMap(({credentials}) => // these are valid here, but cant be passed into the ajax
            Observable
                .ajax.getJSON('http://localhost:3001/login')
                .map((loginResponse) => 
                    loginResponse.status === 200
                        ? loginResolvedAction(loginResponse.response)
                        : loginFailedAction({message: 'Error'})));
Probably going to avoid epics this way until some kind soul can provide documentation or examples.

**Comment by evertbouw** [comment]
@stevematdavies I agree that the docs are lacking, but if you've worked with a different ajax library before you should be able to figure out what is going on with a quick glance at the source code: https://github.com/ReactiveX/rxjs/blob/master/src/observable/dom/AjaxObservable.ts

The `getJSON` method takes a headers object as the second argument. If you need more power than that you can build your own request object and pass that to `ajax({ method: "GET" /* etc */ })`

**Comment by BerkeleyTrue** [comment]
While I still use redux-epic on my legacy projects, redux-observable is now my go to. Combined with Redux-first-router it makes SSR much simpler from the stand point of this library (RFR does the heavy lifting). I'm going to close this issue. 

@stevematdavies I don't see anything wrong with your example from the stand point of the API of Rx or RO. Personally I'm partial to reading the translated source code of Rx to diagnose issues. The source for ajax methods can be found here http://reactivex.io/rxjs/file/es6/observable/dom/AjaxObservable.js.html

Without more information about the exact failure I don't think I can be of more help. Maybe putting together a JSBin will help us help you. 

Happy Coding

**Comment by rgupta2** [comment]
@BerkeleyTrue Could you point to a project using redux-observable and Redux-first-router together. The action I am dispatching via redux-first-router is reaching my reducer but not the epic.


```
const { reducer: routeReducer, middleware: routeMiddleware, enhancer: routeEnhancer } = connectRoutes(
  history,
  routesMap,
  {
    querySerializer: queryString,
    onAfterChange: (dispatch: Dispatch<any>, getState: StateGetter<any>) => {
      const state = getState();      
      const query = state.location && state.location.query && state.location.query.search;
      
      dispatch(fetchPage.started({index: 0,
        pageSize: 10,
        sort: [],
        query }));
    }
  }
); // yes, 3 redux aspects

```
